File: clients/java/src/main/java/com/hook0/client/JsonException.java
package com.hook0.client;

/**
 * The one failure {@link Json} reports, whichever way it was reading or writing.
 *
 * <p>It covers a document that is not JSON, one that crosses a ceiling {@link Json} sets for itself, and a value this
 * client was asked to write that JSON has no way to carry. Nothing else ever leaves that reader: a caller that catches
 * this has caught every way a document can fail to be one.
 */
public final class JsonException extends Hook0Exception {

  private static final long serialVersionUID = 1L;

  /**
   * Builds one out of what could not be read or written.
   *
   * @param detail what to say about the failure
   */
  public JsonException(String detail) {
    super(detail);
  }
}

File: clients/kotlin/src/main/kotlin/com/hook0/kotlin/JsonException.kt
package com.hook0.kotlin

/**
 * A document the reader this artefact carries could not read, or a value it could not write.
 *
 * Every ceiling [Json] applies is reported through this rather than by trimming what crossed it:
 * half a document read as a whole one is worse than no document at all.
 *
 * @param detail what to say about the failure
 */
class JsonException(detail: String) : Hook0Exception(detail)
